File: hackathons/workshops/payment-plans/python/subscriber.py
"""
Nevermined Lab: Payment Plans — Subscriber Flow

The subscriber experience: check balance, buy plan, get token, consume.
"""

import os
import httpx
import urllib3
from payments_py import Payments, PaymentOptions
from dotenv import load_dotenv

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
load_dotenv()

payments = Payments.get_instance(
    PaymentOptions(
        nvm_api_key=os.getenv("NVM_SUBSCRIBER_API_KEY", ""),  # subscriber key
        environment=os.getenv("NVM_ENVIRONMENT", "sandbox"),
    )
)

PLAN_ID = os.getenv("NVM_PLAN_ID", "")
SERVER_URL = "http://localhost:4000"


def main():
    # 1. Check balance
    balance = payments.plans.get_plan_balance(PLAN_ID)
    print(f"Credits remaining: {balance}")

    # 2. Buy plan if needed
    if balance == 0:
        order = payments.plans.order_plan(PLAN_ID)
        print("Plan purchased:", order)

    # 3. Get access token
    token_result = payments.x402.get_x402_access_token(PLAN_ID)
    access_token = token_result["accessToken"]
    print("Token obtained")

    # 4. Use the token
    with httpx.Client(timeout=30.0) as client:
        response = client.post(
            f"{SERVER_URL}/ask",
            headers={"payment-signature": access_token},
            json={"query": "Analyze this data"},
        )
        print("Response:", response.json())


if __name__ == "__main__":
    main()

File: hackathons/workshops/payment-plans/ts/subscriber.ts
/**
 * Nevermined Lab: Payment Plans — Subscriber Flow
 *
 * The subscriber experience: check balance, buy plan, get token, consume.
 */

import { Payments, EnvironmentName } from "@nevermined-io/payments";

const payments = Payments.getInstance({
  nvmApiKey: process.env.NVM_API_KEY!, // subscriber key
  environment: (process.env.NVM_ENVIRONMENT || "sandbox") as EnvironmentName,
});

const PLAN_ID = process.env.NVM_PLAN_ID!;
const SERVER_URL = "http://localhost:3000";

async function main() {
  // 1. Check balance
  const balance = await payments.plans.getPlanBalance(PLAN_ID);
  console.log(`Credits remaining: ${balance.balance}`);

  // 2. Buy plan if needed
  if (balance.balance.toString() === "0") {
    const order = await payments.plans.orderPlan(PLAN_ID);
    console.log("Plan purchased:", order);
  }

  // 3. Get access token
  const { accessToken } = await payments.x402.getX402AccessToken(PLAN_ID);
  console.log("Token obtained");

  // 4. Use the token
  const response = await fetch(`${SERVER_URL}/ask`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "payment-signature": accessToken,
    },
    body: JSON.stringify({ query: "Analyze this data" }),
  });

  console.log("Response:", await response.json());
}

main().catch(console.error);
